src/interpret.cpp

```diff
@@ -292,11 +298,62 @@ static inline bool csr_is_read_only(CSR_address csraddr) {
     return ((to_underlying(csraddr) & 0xc00) == 0xc00);
 }
 
-/// \brief Extract privilege level from CSR address.
+/// \brief Checks if CSR access is allowed from the current mode.
+/// \param a Machine state accessor object.
 /// \param csr Address of CSR in file.
-/// \returns Privilege level.
-static inline uint32_t csr_priv(CSR_address csr) {
-    return (to_underlying(csr) >> 8) & 3;
+/// \param status cause code in case of failure.
+/// \returns true if CSR access is allowed from the current mode, false otherwise.
+template <typename STATE_ACCESS>
+static inline bool csr_is_allowed_access(STATE_ACCESS &a, CSR_address csr, MCAUSE_constants &status) {
+    uint8_t cur_priv = a.read_iflags_PRV();
+    uint8_t access_priv = (cur_priv == PRV_HS && !is_virtual_mode(a)) ? 2 : cur_priv;
+    uint8_t register_priv = (to_underlying(csr) >> 8) & 3;
+
+    if (access_priv < register_priv) {
+        // we should not access Hypervisor and Virtual registers directly from the virtual mode
+        // for Hypervisor and Virtual registers register_priv = 2
+        if (is_virtual_mode(a) && register_priv <= PRV_HS + 1)
+            status = MCAUSE_VIRTUAL_INSTRUCTION;
+        else status = MCAUSE_ILLEGAL_INSN;
+        return false;
+    }
+
+    return true;
+}
+
+/// \brief Checks if machine is running in a virtual mode.
+/// \param a Machine state accessor object.
+/// \returns true if machine is running in a virtual mode, false otherwise.
+template <typename STATE_ACCESS>
+static inline uint64_t is_virtual_mode(STATE_ACCESS &a) {
+    if (!(a.read_misa() & MISA_EXT_H_MASK)) { // hypervisor extension disabled
+        return false;
+    }
+    return a.read_iflags_VRT();
+}
+
+static inline uint8_t encode_access_mode(uint8_t priv, uint8_t virt) {
+    uint8_t access_mode = (virt << 7) | priv;
```

Comment:
I am starting to think we could have an mode enum with 3 bits with the follwing layout:
```
| 1   |  2  |
| VRT | PRV |
```

Then
```c++
enum PRV_constants : uint8_t {
    PRV_U = 0,    ///< User mode
    PRV_S = 1,    ///< Supervisor mode
    PRV_RSVD = 2, ///< Reserved
    PRV_M = 3     ///< Machine mode
};

enum MODE_shifts : uint8_t {
    ACCESS_MODE_PRV_SHIFT = 0,
    ACCESS_MODE_VRT_SHIFT = 2
};

enum MODE_masks : uint8_t {
    ACCESS_MODE_PRV_MASK = 3 << ACCESS_MODE_PRV_SHIFT,
    ACCESS_MODE_VRT_MASK = 1 << ACCESS_MODE_VRT_SHIFT
};

enum MODE_constants : uint8_t {
  MODE_U = PRV_U,
  MODE_HS = PRV_S,
  MODE_M = PRV_M,
  MODE_VS = ACCESS_MODE_VRT_MASK | PRV_S,
  MODE_VU = ACCESS_MODE_VRT_MASK | PRV_U
};
```

Since MODE is a combination of PRV and VRT,  we could then clarify some portions of the code in this file using this enum, instead of many `is_virtual_mode(a) && prv == PRV_HS` checks, it would make things more readable.